src/attachments/imageResizer.js

```diff
@@ -1,118 +1,37 @@
-import fs from "fs-extra";
-import jimp from "jimp";
-import imagemin from "imagemin";
-import pngquant from "imagemin-pngquant";
-import jpegopti from "imagemin-jpegoptim";
+import sharp from "sharp";
 
 export function generateThumb(options) {
-  const {fromPath, toPath, imageHeight, imageWidth, minify} = options;
-  return readImage(fromPath)
-    .then(resizeImage({
-      imageHeight,
-      imageWidth
-    }))
-    .then(minifyImage(minify))
-    .then(saveImage(toPath));
+  const { fromPath, toPath, imageHeight, imageWidth, minify = false } = options;
+
+  return getTransformer({ fromPath, resize: true, imageHeight, imageWidth, minify }).then(sharp => sharp.toFile(toPath));
 }
 
 export function reduceImage(options) {
-  const {fromPath, toPath} = options;
-  return readImage(fromPath)
-    .then(image => {
-      return new Promise((resolve, reject) => {
-        image.getBuffer(image._originalMime, (err, buffer) => {
-          if (err) {
-            console.log(`could not get buffer to copy with mime type ${image._originalMime}`);
-            reject(err);
-          } else {
-            resolve(buffer);
-          }
-        });
-      });
-    })
-    .then(minifyImage(true))
-    .then(saveImage(toPath));
-}
+  const { fromPath, toPath } = options;
 
-function readImage(fromPath) {
-  return new Promise((resolve, reject) => {
-    jimp.read(fromPath, (err, image) => {
-      if (err) {
-        console.log(`Could not read ${fromPath}`);
-        reject(err);
-      } else {
-        resolve(image);
-      }
-    });
-  });
+  return getTransformer({ fromPath, resize: false, minify: true }).then(sharp => sharp.toFile(toPath));
 }
 
-function resizeImage(
-  {
-    imageHeight = "auto",
-    imageWidth = "auto"
-  }) {
-
-  const height = imageHeight === "auto" ? jimp.AUTO : imageHeight;
-  const width = imageWidth === "auto" ? jimp.AUTO : imageWidth;
-  const isFittingToScale = imageHeight !== "auto" && imageWidth !== "auto";
-  return image => new Promise((resolve, reject) => {
-    // resize the width to <imageSize> and scale the height accordingly
+function getTransformer({ fromPath, resize, imageHeight, imageWidth, minify }) {
+  return new Promise((resolve, reject) => {
     try {
-      const bufferCb = (err, buffer) => {
-        if (err) {
-          console.log(`could not get buffer to resize with mime type ${image._originalMime}`);
-          reject(err);
-        } else {
-          resolve(buffer);
-        }
-      };
-      if (isFittingToScale) {
-        image
-          .contain(width, height)
-          .getBuffer(image._originalMime, bufferCb);
-      } else {
-        image
-          .resize(width, height)
-          .getBuffer(image._originalMime, bufferCb);
+      const transformer = sharp(fromPath);
+
+      if (resize) {
+        transformer.resize({
```

Comment:
[question] Verstehe ich nicht. Ist der Transformer jetzt eine Pipeline (dann wäre das falsch, wird aber in der Doku als "Pipeline" bezeichnet) oder stateful (dann wäre das vielleicht korrekt).

Reply:
Ich verstehe die Doku so, dass man beides machen kann:

> Constructor factory to create an instance of sharp, to which further methods are chained.
> 
> JPEG, PNG, WebP, GIF, AVIF or TIFF format image data can be streamed out from this object. When using Stream based output, derived attributes are available from the info event.

Ein Beispiel aus der Doku:

```
var transformer = sharp()
  .resize(300)
  .on('info', function(info) {
    console.log('Image height is ' + info.height);
  });
readableStream.pipe(transformer).pipe(writableStream);
```
Das streamed dann wohl das Ganze.

Ein anderes Beispiel:
```
// Read a raw array of pixels and save it to a png
const input = Uint8Array.from([255, 255, 255, 0, 0, 0]); // or Uint8ClampedArray
const image = sharp(input, {
  // because the input does not contain its dimensions or how many channels it has
  // we need to specify it in the constructor options
  raw: {
    width: 2,
    height: 1,
    channels: 3
  }
});
await image.toFile('my-two-pixels.png');
````
Hier werden einfach die Funktionen gechained.